securepreferences

sashei opened the pull request:
resolves #67 


remaining work:
 - [x] updating the tests for `SecurePreferences` to properly mock the Keystore (conversation in a-c is here: mozilla-mobile/android-components#1072)
 - [x] seeing how this could be integrated with the work on the Fingerprint Manager (issue #68) using the Keystore helpers (waiting for @ioana-farcas to push work there)
 - [x] finishing up AccountStore class with the full FxA flow & proper unit tests (addressed in issue #22 but some work still required there. also difficult to test because we cannot run unit tests on the cross-compiled rust code. application-services issue here: mozilla/application-services#262)

linuxwolf commented on app/src/main/java/mozilla/lockbox/store/SecureStore.kt:
```diff
@@ -0,0 +1,97 @@
+            .subscribe {
+                when (it) {
+                    is SecureAction.OAuthInfo -> {
+                        // tbd: how to convert to & from String values for OAuthInfo...
```
yeah yeah yeah, WIP (-:

This might provide some pointers for the `tbd`s: [samples/firefox-accounts](https://github.com/mozilla-mobile/android-components/blob/dbaf87110dbf5f7183ba067c086049f4b4ff5bb6/samples/firefox-accounts/src/main/java/org/mozilla/samples/fxa/MainActivity.kt#L137)

sashei replied:
thanks matt :)

jhugman commented on app/src/main/java/mozilla/lockbox/support/Optional.kt:
```diff
@@ -0,0 +1,10 @@
+package mozilla.lockbox.support
+
+data class Optional<T>(val value: T?)
+fun <T> T?.asOptional() = Optional(this)
```
Ooh. Weird.

sashei replied:
gotta make Rx play nicely with optionals...

jhugman replied:
Yeah, that sucks. Perhaps we shouldn't be using optionals in Rx?

Playing around, there don't seem to be any good answers to this.

sashei replied:
The concept of optionals in our observables is useful; we use the presence / absence of things like OAuthInfo to do routing in iOS and I'm trying to think of how it would look otherwise:

We could be more _explicit_ by having something like `val accountAvailable: Observable<Bool>` but it's another piece of state to update along with `oauthInfo` / `profile`. 

To avoid updating that state every time, we could just implement a filter for `oauthInfo == null`, which brings us back to the original problem of not being able to represent Optionals in Rx. I think we'll want them in other places so I'm OK with this solution for now.

At its core, this is a workaround for RxKotlin being implemented in Java and Java not having a concept of Optionals.

sashei commented on app/src/main/java/mozilla/lockbox/action/SecureAction.kt:
```diff
@@ -0,0 +1,14 @@
+
+sealed class SecureAction : Action {
+    data class FxAAccount(val account: FirefoxAccount) : SecureAction()
+}
```
I guess I was thinking you would "secure" your FxA account as an action, but yeah I can see the ambiguity.

sashei commented:
I realized in addressing these comments that there's been some scope drift in this PR. I've converted the `SecureStore` to an `AccountStore` to match iOS, and made a `SecurePreferences` helper class that actually addresses the ticket.

There's some remaining work, which includes:
- updating the tests to properly mock the `Keystore` and test the `SecurePreferences` class (conversation in `a-c` is here: https://github.com/mozilla-mobile/android-components/issues/1072)
- seeing how this could be integrated with the work on the Fingerprint Manager (issue #68) using the `Keystore` helpers (waiting for @ioana-farcas to push work there)
- finishing up `AccountStore` class with the full FxA flow & proper unit tests (addressed in issue #22 but some work still required there. also difficult to test because we cannot run unit tests on the cross-compiled rust code. `application-services` issue here: https://github.com/mozilla/application-services/issues/262)

sashei commented:
follow-up: we still need to make `Keystore` open because you cannot do the `mock-maker-inline` solution for instrumentation tests, and with the version of `Base64` from the examples, you cannot run this code _outside_ of an Android VM in unit tests D:

jhugman commented on app/src/main/java/mozilla/lockbox/store/AccountStore.kt:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    fun apply(sharedPreferences: SharedPreferences) {
+        securePreferences.apply(sharedPreferences)
```
Pass in the `SecurePreferences` object, not the `SharedPreferences` object.

sashei replied:
That means `LockboxApplication` will be doing heavier lifting than before, but I suppose that's fine.

linuxwolf replied:
put me down for "not a fan of this change", unless `SecurePreferences` is maintained/initialized by `LockboxApplication` for other pieces.

sashei replied:
:sparkles: _no one knows_ :sparkles:

sashei replied:
I don't like any of the above tbh; I think we agreed to keep the `apply` pattern to get those other PRs landed, but I would rather come up with another solution because there will undoubtedly be many other `Context`-requiring `Store`s.

jhugman commented on app/src/main/java/mozilla/lockbox/store/AccountStore.kt:
```diff
@@ -0,0 +1,75 @@
+        loadFxA()
+    }
+
+    private fun loadFxA() {
```
Nit: there's a bunch of things I don't like about this naming. 

First: prefer `FirefoxAccount` or `Account` to `FxA`.

Second: you're not returning anything (as you might expect to `load` an existing account); you're then resaving immediately after you've loaded, or updating the profile/oauth with `null`.

sashei replied:
We need to resave immediately after loading the `FirefoxAccount` from the JSON string because that's when they do the account refresh.

I'm mirroring what we have in iOS here, but I'm happy to refactor if it would be clearer.

jhugman commented on app/src/main/java/mozilla/lockbox/store/AccountStore.kt:
```diff
@@ -0,0 +1,75 @@
+        val oauthSubject = oauthInfo as PublishSubject
+        account.getOAuthToken(FXA_SCOPES).whenComplete {
+            oauthSubject.onNext(it.asOptional())
+        }
```
Similarly, this method needs a rethink/rename. 

We'e not just `loading` or `persisting` in either of these methods. 

Perhaps they need to be split up into smaller methods.

sashei replied:
Maybe we need the rename for `persist`. We should never save the `FirefoxAccount` without updating the `oauthInfo` and `profile` observables to reflect the saved state, so they should logically be grouped together in some way.

sashei replied:
I also think of `load` and `persist` as extending to loading / persisting things into our external `Observable`s.

jhugman commented on app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt:
```diff
@@ -0,0 +1,46 @@
+        }
+    }
+
+    open fun putString(key: String, value: String) {
```
Nit: try and make this mirror the `SharedPreferences` API as much as possible.

sashei replied:
This is functionally the same, but I felt like the "editor" concept was pretty high overhead for a two method class.

jhugman replied:
👍 

Eventually, I think we should try and make this a version for `android-components` that is a drop in replacement/wrapper for `SharedPreferences`.

sashei commented:
@jhugman I don't love the `apply` pattern either. What other ideas did we have to get context in stores where it's necessary?

linuxwolf commented on app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt:
```diff
@@ -0,0 +1,46 @@
+private const val KEYSTORE_LABEL = "lockbox-keystore"
+private const val BASE_64_FLAGS = Base64.URL_SAFE or Base64.NO_PADDING
+
+open class SecurePreferences(
```
At some point, `Keystore.generateKey()` must be called, or all the calls to encrypt/decrypt will error.

sashei replied:
oh.

linuxwolf replied:
my slight bias is to do this lazily, at the point where encrypt/decrypt are first called:

* the provisioning is cheap, so the latency is minor
* it's arguably a better place in the app's lifecycle to deal with errors?
* If it doesn't exist when there's definitely a value to decrypt, we can differentiate the failure better (e.g., _something_ about the app left `SharedPreferences` around but not an encryption key, vs the decrypt failed because the keys are different)?

sashei replied:
Ok, I'd like to link up with you to talk through the error cases and what handling we should have here.

linuxwolf commented:
> @jhugman I don't love the `apply` pattern either. What other ideas did we have to get context in stores where it's necessary?

Are we discussing this _again_? (-:

sashei commented on app/src/main/java/mozilla/lockbox/store/AccountStore.kt:
```diff
@@ -0,0 +1,75 @@
+import mozilla.lockbox.support.SecurePreferences
+
+private const val FIREFOX_ACCOUNT_KEY = "firefox-account"
+private val FXA_SCOPES = arrayOf("profile", "https://identity.mozilla.com/apps/lockbox")
```
why do we need oldsync for Android? I was confused about this when I saw your bugzilla conversation go by. I thought we only needed oldsync in iOS for the FxiOS Sync code and the application-services version will work with the new `sync`

linuxwolf replied:
That is not correct; the `oldsync` scope gets us the FxSync key.  Even the new `sync` implementation needs this same key, or nothing from FxSync will decrypt.

jhugman commented on app/src/test/java/mozilla/lockbox/support/SecurePreferencesTest.kt:
```diff
@@ -0,0 +1,163 @@
+import org.junit.runner.RunWith
+import org.mockito.Mock
+import org.mockito.Mockito
+import org.mockito.Mockito.`when` as whenCalled
```
Ooh, fancy `as`.

jhugman commented on app/src/test/java/mozilla/lockbox/store/AccountStoreTest.kt:
```diff
@@ -0,0 +1,33 @@
+//        profileObserver.assertValue(Optional<Profile>(null))
+//    }
+
+    // note: further FxA-related tests on hold until we can use x-compiled Rust code in unit specs
```
Ok.

jhugman commented on app/src/test/java/mozilla/lockbox/ExampleUnitTest.kt:
```diff
@@ -1,23 +0,0 @@
-    fun addition_isCorrect() {
-        assertEquals(4, 2 + 2)
-    }
-}
```
👍 All the thumbs. I hadn't realised this was still here.

jhugman commented on app/src/main/java/mozilla/lockbox/store/AccountStore.kt:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+
+    private fun pushNullAndReset() {
```
Note: I hate names with conjunctions in. Please don't fix this in this PR.
